planner/core/runtime_filter_generator.go

```diff
@@ -0,0 +1,250 @@
+	columnUniqueIdToRF            map[int64][]*RuntimeFilter
+	parentPhysicalPlan            PhysicalPlan
+	childIdxForParentPhysicalPlan int
+}
+
+// GenerateRuntimeFilter is the root method.
+// It traverses the entire tree in preorder.
+// It constructs RF when encountering hash join, and allocate RF when encountering table scan.
+// It realizes the planning of RF in the entire plan tree.
+// For example:
+/*
+PhysicalPlanTree:
+      HashJoin
+       /     \
+ TableScan   ExchangeNode
+
+1. generateRuntimeFilter
+      HashJoin (with RF1)
+       /     \
+ TableScan   ExchangeNode
+
+2. assignRuntimeFilter
+      HashJoin (with RF1)
+       /     \
+ TableScan   ExchangeNode
+(assign RF1)
+*/
+func (generator *RuntimeFilterGenerator) GenerateRuntimeFilter(plan PhysicalPlan) {
+	switch physicalPlan := plan.(type) {
+	case *PhysicalHashJoin:
+		generator.generateRuntimeFilterInterval(physicalPlan)
+	case *PhysicalTableScan:
+		generator.assignRuntimeFilter(physicalPlan)
+	case *PhysicalTableReader:
+		generator.parentPhysicalPlan = plan
+		generator.childIdxForParentPhysicalPlan = 0
+		generator.GenerateRuntimeFilter(physicalPlan.tablePlan)
+		physicalTableReader := plan.(*PhysicalTableReader)
+		if physicalTableReader.StoreType == kv.TiFlash {
+			physicalTableReader.TablePlans = flattenPushDownPlan(physicalTableReader.tablePlan)
+		}
+	}
+
+	for i, children := range plan.Children() {
+		generator.parentPhysicalPlan = plan
+		generator.childIdxForParentPhysicalPlan = i
+		generator.GenerateRuntimeFilter(children)
+	}
+}
+
+func (generator *RuntimeFilterGenerator) generateRuntimeFilterInterval(plan PhysicalPlan) {
+	hashJoinPlan := plan.(*PhysicalHashJoin)
+	// precondition: the storage type of hash join must be TiFlash
+	if hashJoinPlan.storeTp != kv.TiFlash {
+		logutil.BgLogger().Warn("RF only support TiFlash compute engine while storage type of hash join node is not TiFlash",
+			zap.Int("PhysicalHashJoinId", hashJoinPlan.id),
+			zap.String("StoreTP", hashJoinPlan.storeTp.Name()))
+		return
+	}
+	// check hash join pattern
+	if !generator.matchRFJoinType(hashJoinPlan) {
+		return
+	}
+	// check eq predicate pattern
+	for _, eqPredicate := range hashJoinPlan.EqualConditions {
+		if generator.matchEQPredicate(eqPredicate, hashJoinPlan.RightIsBuildSide()) {
+			// construct runtime filter
+			newRFList, targetColumnUniqueId := NewRuntimeFilter(generator.rfIdGenerator, eqPredicate, hashJoinPlan)
+			// update generator rf list
+			rfList := generator.columnUniqueIdToRF[targetColumnUniqueId]
+			if rfList == nil {
+				generator.columnUniqueIdToRF[targetColumnUniqueId] = newRFList
+			} else {
+				generator.columnUniqueIdToRF[targetColumnUniqueId] = append(generator.columnUniqueIdToRF[targetColumnUniqueId], newRFList...)
+			}
+		}
+	}
+}
+
+func (generator *RuntimeFilterGenerator) assignRuntimeFilter(plan PhysicalPlan) {
+	// only full scan
+	physicalTableScan := plan.(*PhysicalTableScan)
+	if !physicalTableScan.isFullScan() {
+		return
+	}
```

Comment:
Is this a necessary check, or just in case something unexpected happens?

Reply:
It is a necessary check. If table scan is not a full scan, the runtime filter will not be assign

Reply:
I don't quite understand this. Would it affect correctness?

Reply:
It is my fault. The runtime filter could be assign whatever fullscan or not.